daemon/cmd/ipam.go

```diff
@@ -323,28 +335,37 @@ func (d *Daemon) allocateHealthIPs() error {
 				result.CIDRs = coalesceCIDRs(result.CIDRs)
 			}
 
-			log.Debugf("IPv4 health endpoint address: %s", result.IP)
+			log.Infof("IPv4 health endpoint address: %s", result.IP)
 			node.SetEndpointHealthIPv4(result.IP)
 
 			// In ENI and AlibabaCloud ENI mode, we require the gateway, CIDRs, and the ENI MAC addr
 			// in order to set up rules and routes on the local node to direct
 			// endpoint traffic out of the ENIs.
 			if option.Config.IPAM == ipamOption.IPAMENI || option.Config.IPAM == ipamOption.IPAMAlibabaCloud {
-				var err error
 				if d.healthEndpointRouting, err = parseRoutingInfo(result); err != nil {
 					log.WithError(err).Warn("Unable to allocate health information for ENI")
 				}
 			}
 		}
 
 		if option.Config.EnableIPv6 {
-			result, err := d.ipam.AllocateNextFamilyWithoutSyncUpstream(ipam.IPv6, "health", ipam.PoolDefault())
-			if err != nil {
-				if healthIPv4 := node.GetEndpointHealthIPv4(); healthIPv4 != nil {
-					d.ipam.ReleaseIP(healthIPv4, ipam.PoolDefault())
-					node.SetEndpointHealthIPv4(nil)
+			healthIP := node.GetEndpointHealthIPv6()
+			if healthIP != nil {
+				result, err = d.ipam.AllocateIPWithoutSyncUpstream(healthIP, "health", ipam.PoolDefault())
+				if err != nil {
+					log.Warn("unable to re-allocate health IPv6.")
+					healthIP = nil
+				}
+			}
+			if healthIP == nil {
+				result, err = d.ipam.AllocateNextFamilyWithoutSyncUpstream(ipam.IPv6, "health", ipam.PoolDefault())
+				if err != nil {
+					if healthIPv4 := node.GetEndpointHealthIPv4(); healthIPv4 != nil {
```

Comment:
Do we need to refetch or could be just check if healthIpv4 was set above and use it?

Reply:
Done